oat-sa/tao-core-ui-fe
# Fix/tr 1460/better detection of media disconnection

Related to: https://oat-sa.atlassian.net/browse/TR-1460

Improve the detection of stalled media in the media player.

The previous way of detecting the stalled media has been removed and totally reworked. Instead of having it both in the mediaplayer and in the html5 implementation, it has been entirely integrated in the html5 implementation since it is deeply linked to it.

The stalled state is detected at various stages:
- when an explicit error occurs
- when the playback is stuck for some time (2s by default)
- when the current position is suddenly moved to the end (browser handling of missing data)

For this purpose, several counter measures have been put in place:
- a timeObserver to detect irregularities in the timeline
- a reminder to detect stuck playback (a callback is registered and is called if the time position didn't get updated for more than N seconds)
- an error observer that focus on recoverable errors and discard those that are not impacting the playback

**How to test:**
- have a video with a size big enough to not be entirely cached when starting the playback (40~60MB video would be enough).
- disconnect the network and play the video
- disconnect the network and reload the player
- play the video, disconnect the network and change the time position
- when the connection is off and the video is playing back, the message "You are encountering a prolonged connectivity loss. Click to reload." Should be displayed and the video should be blurred.

**Note about how to disconnect from the network. It can be done either by:**
- set the browser offline using the dev tools (Chrome, Edge) or by a dedicated menu (Firefox)
- block the request from the dev tools
- shutdown the server
- put an error in the file `tao/getFileFlysystem.php`, like `header("HTTP/1.0 404 Not Found"); die();`

**Note about unit tests:**
- unit tests need to be updated, and they are failing for the time being. It should be fixed before merging the PR
